OSOE-609: Add fake video capture sample test in Lombiq.UITestingToolbox

dministro opened the pull request:
[OSOE-609](https://lombiq.atlassian.net/browse/OSOE-609)

[OSOE-609]: https://lombiq.atlassian.net/browse/OSOE-609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dministro committed:
Updating submodules:
 - Lombiq.ContentEditors: fast-forward
 - Lombiq.VueJs: Adding sample application and recipe

dministro committed:
Allowing words
Updating submodule:
 - Lombiq.VueJs: Fix spelling

dministro committed:
Adding UI tests
Updating submodule:
 - Lombiq.VueJs: Adding UI tests

dministro committed:
Excluding files
Allowing words

dministro committed:
Fixing excludes.txt

dministro committed:
dummy

dministro committed:
Adding coments
Updating submodules:
 - Lombiq.UITestingToolbox: Adding docs
 - Lombiq.VueJs: Adding comments

dministro committed:
Updating submodule:
 - Lombiq.UITestingToolbox: Fixing double spaces

dministro committed:
Updating submodule:
 - Lombiq.UITestingToolbox: fast-forward

0liver commented on .github/actions/spelling/excludes.txt:
```diff
@@ -1,3 +1,4 @@
+Lombiq\.VueJs\.Tests\.UI\/Assets\/Media\/(john\.doe|richard\.roe).mjpeg$
```
Let's simplify and make this more robust:
```suggestion
Lombiq\.VueJs\.Tests\.UI/Assets/Media/*.mjpeg$
```

0liver commented on test/Lombiq.OSOCE.Tests.UI/Tests/ModuleTests/BehaviorVueTests.cs:
```diff
@@ -24,4 +29,44 @@ public BehaviorVueTests(ITestOutputHelper testOutputHelper)
+    {
+        var errors = (await validationResult.GetErrorsAsync())
+            .Where(error =>
+                !error.ContainsOrdinalIgnoreCase("The autoplay attribute is not allowed on <video>")
+                && !error.ContainsOrdinalIgnoreCase("title text cannot be longer than 70 characters"));
```
Where do these errors originate for you?

dministro replied:
These are html validator errors:
```
error: The autoplay attribute is not allowed on <video> (no-autoplay) at e3744494-c9ae-43b7-8e9e-b4d6ac04d82b.html:331:247:
  329 |     </div>
  330 |     <loading-indicator v-else-if="loading" /></script>
> 331 | <div id="qrCardApp"><div class="container"><div class="row justify-content-center mb-3"><div class="col-md-6 col-sm"><div class="card"><div class="card-body"><!----> <div data-v-35411cc1="" class="qrcode-stream-wrapper"><video data-v-35411cc1="" autoplay="autoplay" muted="muted" playsinline="" class="qrcode-stream-camera qrcode-stream-camera--hidden"></video><canvas data-v-35411cc1="" class="qrcode-stream-camera"></canvas><canvas data-v-35411cc1="" class="qrcode-stream-overlay"></canvas><div data-v-35411cc1="" class="qrcode-stream-overlay"><blockquote class="blockquote text-center" data-v-35411cc1="">
      |                                                                                                                                                                                                                                                       ^^^^^^^^^^^^^^^^^^^
  332 |         Loading...
  333 |     </blockquote></div></div></div></div></div></div> <!----> <!----></div></div>
  334 |
Details: https://html-validate.org/rules/no-autoplay.html


1 error found.
```

```
error: title text cannot be longer than 70 characters (long-title) at f3d51ba8-d411-469e-8361-55c3415501c2.html:11:42:
   9 | IsCached: False
  10 | --&gt;
> 11 | Blog - Lombiq's OSOCE - UI Testing</title>
     |                                          ^
  12 |     <link type="image/x-icon" rel="shortcut icon" href="/favicon.ico">
  13 |
  14 |     <script src="/Lombiq.BaseTheme/vendors/bootstrap/js/bootstrap.bundle.js?v=SIYbLg8Kyb12W8RgxT7zwNAeUMfrfld1XpKbc_0Q1hE"></script>
Details: https://html-validate.org/rules/long-title.html


1 error found.
```

The second one is a `Shape Tracing` issue, if I remember well.

0liver replied:
I've made a note on this under the issue: https://github.com/Lombiq/Helpful-Extensions/issues/120#issuecomment-1517600934.

0liver commented on test/Lombiq.OSOCE.Tests.UI/Tests/ModuleTests/BehaviorVueTests.cs:
```diff
@@ -24,4 +29,44 @@ public BehaviorVueTests(ITestOutputHelper testOutputHelper)
+                configuration.BrowserConfiguration.ConfigureFakeVideoSourceForNegativeTest();
+                configuration.HtmlValidationConfiguration.AssertHtmlValidationResultAsync = AssertHtmValidationResultAsync;
+                configuration.AssertBrowserLog = logEntries =>
+                    OrchardCoreUITestExecutorConfiguration.AssertBrowserLogIsEmpty(
+                        logEntries.Where(logEntry =>
+                            // The fetch call reports an error to the browser console if the QrCardController.GetBusinessCard
+                            // action results in NotFound.
+                            !(
+                                logEntry.Message.ContainsOrdinalIgnoreCase(nameof(QrCardController.GetBusinessCard))
+                                && logEntry.Message.ContainsOrdinalIgnoreCase(
+                                    "Failed to load resource: the server responded with a status of 404"))));
```
To me this looks like your asserting that the browser log does not contain the given messages. I would expect the test to assert that they **do** exist. What gives?

dministro replied:
Agree, but the same validation is used for validating logs eg. after the setup phase, so I can't expect something here that is not happening in the setup phase.

0liver replied:
My point is that this code **filters** the messages that should exist in the error case. You shouldn't assert that the browser log is empty, since obviously it won't be because you're generating an error on purpose.

Instead of this code, you should assert that the expected error messages **do exist** in the browser console.

dministro replied:
If you start the test in debug mode, and place the breakpoint to line 54, you can see that, the same assertion method gets called on every page change:

Setup page:
![image](https://user-images.githubusercontent.com/15349881/233614400-df46a849-dc18-438c-8a59-8e6df249fa00.png)

After setup:
![image](https://user-images.githubusercontent.com/15349881/233614555-5a6b9b3e-4116-4052-bb5f-6117d9755470.png)

After the test completed:
![image](https://user-images.githubusercontent.com/15349881/233615045-1e80c0b6-8b1d-4f44-95c2-dd9d8853594f.png)

So, how can I expect that the error message **do exist** only at the end of the test, while in every other case of assertion, the "expected" message doesn't exist?

0liver replied:
Now I understand. Thanks for the explanation.

dministro committed:
Apply suggestions from code review

[no ci]

Co-authored-by: 0liver <0liver@users.noreply.github.com>

dministro committed:
Updating submodules:
 - Lombiq.VueJs: Addressing "In that case, do the same in ContentPart.Edit-BusinessCard.cshtml, too, please." ...
 - Lombiq.UITestingToolbox: Update Lombiq.Tests.UI/Docs/FakeVideoCaptureSource.md

dministro committed:
Updating submodule:
 - Lombiq.VueJs: Using Lazy

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-609

dministro committed:
Updating submodules

dministro committed:
Updating submodule:
 - Lombiq.VueJs: Localizing strings in template

dministro committed:
Updating submodule:
 - Lombiq.VueJs: Removing message property of loading-indicator

dministro committed:
Updating submodule:
 - Lombiq.VueJs: Fixing analizer violation

Commit:
Format comment to 120 chars

Commit:
Update vue component formatting

[skip ci]

dministro committed:
Updating submodule:
 - Lombiq.VueJs: Adjusting UI test

dministro committed:
Updating submodule:
 - Lombiq.VueJs: Better error styling

dministro committed:
Adjusting excludes.txt
Updating submodule:
 - Lombiq.VueJs: Adjusting UI test

dministro commented on .github/actions/spelling/excludes.txt:
```diff
@@ -1,3 +1,4 @@
+Lombiq\.VueJs\.Tests\.UI/Assets/Media/.*\.mjpeg$
```
Partially reverted to fix
![image](https://user-images.githubusercontent.com/15349881/233973267-273e8703-4772-4b81-9649-f9cff8c5b816.png)

Commit:
Update to merged submodules
